skimage/segmentation/random_walker_segmentation.py

```diff
@@ -77,163 +74,195 @@ def _make_graph_edges_3d(n_x, n_y, n_z):
         Graph edges with each column describing a node-id pair.
     """
     vertices = np.arange(n_x * n_y * n_z).reshape((n_x, n_y, n_z))
-    edges_deep = np.vstack((vertices[:, :, :-1].ravel(),
-                            vertices[:, :, 1:].ravel()))
+    edges_deep = np.vstack((vertices[..., :-1].ravel(),
+                            vertices[..., 1:].ravel()))
     edges_right = np.vstack((vertices[:, :-1].ravel(),
                              vertices[:, 1:].ravel()))
     edges_down = np.vstack((vertices[:-1].ravel(), vertices[1:].ravel()))
     edges = np.hstack((edges_deep, edges_right, edges_down))
     return edges
 
 
-def _compute_weights_3d(data, spacing, beta=130, eps=1.e-6,
-                        multichannel=False):
+def _compute_weights_3d(data, spacing, beta, eps, multichannel):
     # Weight calculation is main difference in multispectral version
     # Original gradient**2 replaced with sum of gradients ** 2
-    gradients = 0
-    for channel in range(0, data.shape[-1]):
-        gradients += _compute_gradients_3d(data[..., channel],
-                                           spacing) ** 2
+    gradients = np.concatenate(
+        [np.diff(data[..., 0], axis=ax).ravel() / spacing[ax]
+         for ax in [2, 1, 0] if data.shape[ax] > 1], axis=0) ** 2
+    for channel in range(1, data.shape[-1]):
+        gradients += np.concatenate(
+            [np.diff(data[..., channel], axis=ax).ravel() / spacing[ax]
+             for ax in [2, 1, 0] if data.shape[ax] > 1], axis=0) ** 2
+
     # All channels considered together in this standard deviation
-    beta /= 10 * data.std()
+    scale_factor = -beta / (10 * data.std())
```

Comment:
I don't really understand/like this hard coded multiplication by 10, but it is in the actual implementation...

Reply:
Yeah it's quite bad :-(. (it's likely I was the one to introduce this :-()